portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java

```diff
@@ -86,7 +90,30 @@ public void upgrade() throws UpgradeException {
 
 		String message = "Completed upgrade process ";
 
-		try (Connection connection = getConnection()) {
+		boolean partitionConcurrencyEnabled = false;
+
+		if (GetterUtil.getBoolean(
+				PropsUtil.get("database.partition.enabled")) &&
+			GetterUtil.getBoolean(
+				PropsUtil.get("database.partition.thread.pool.enabled"),
+				true)) {
+
+			partitionConcurrencyEnabled = true;
+		}
+
+		Connection connection = null;
+
+		try {
+			if (partitionConcurrencyEnabled) {
+				connection = (Connection)ProxyUtil.newProxyInstance(
+					ClassLoader.getSystemClassLoader(),
+					new Class<?>[] {Connection.class},
+					new ConnectionThreadProxyHandler());
+			}
+			else {
+				connection = getConnection();
+			}
```

Comment:
@LuisOrtiz89 this section looks problematic. If you see the body of "getConnection()", it is doing special handling for external datasource, but once you go into the parallel branch, it always uses portal's default datasource.
All the logic for connection getting (including possible paralllel proxy wrapping) should be moved into "getConnection()" method, and the method should be promoted to BaseDBProcess in order to share with VerifyProcess.
Make sure the external datasources are properly handled even with parallel activation.